add return_and_correct_aliasing() util for wrapper subclasses

bdhirsh opened the pull request:
This PR adds a `return_and_correct_aliasing()` utility, that wrapper subclasses can use to get correct aliasing. I updated `TwoTensor` to use it, and added some testing that the aliasing of my `TwoTensor` subclass now matches the aliasing behavior of normal tensors.

Right now my test just uses a few hand-picked opinfos (that have varying aliasing behavior). I thought all op infos might be overkill (does that take a while to run?), but I'm happy to add them all if people prefer.

One more general question about this PR: eventually, proper aliasing will be a **requirement** in order for AOTAutograd to handle aliasing/mutations on subclasses properly during compilation. How can we make sure that wrapper subclasses use this API? A few options (from talking to Richard):

(1) Yolo require subclasses to use the API and hope users do as well (what this PR does)

(2) Yolo require subclasses to use the API, but add a kwarg to `_make_wrapper_subclass`, e.g. `manual_aliasing=True`, that torch.compile checks for before allowing the subclass to be used in compilation

(3) Automatically run this API in our python fallback, for **every** tensor subclass that currently implements `__tensor_flatten__` (aka only the "traceable" subclasses)

(4) Automatically run this API in our python fallback, for **every** tensor subclass. This would be a bit higher blast radius, since it would change the existing aliasing behavior of wrapper subclasses. Maybe.. this is the right thing to do though?

Either way, my tentative plan is to do (1) to unblock, and revisit this later once we want to come up with public docs + a more general "tensor subclass in PT2 requirements" plan

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* #106406
* #107917
* #106404
* #107926
* #107916
* __->__ #107915

bdhirsh committed:
add return_and_correct_aliasing() util for wrapper subclasses

[ghstack-poisoned]

bdhirsh commented on torch/utils/_python_dispatch.py:
```diff
@@ -165,3 +165,130 @@ def transform_subclass(t, callback):
+            try:
+                torch.ops.aten.set_.source_Storage_storage_offset(ret, arg.untyped_storage(), 0, ret.shape)
+            finally:
+                torch._C._set_meta_in_tls_dispatch_include(meta_in_tls)
```
This is a bit sketchy: I want to slam the output storage onto the input (to guarantee correct aliasing), but:

(1) I use `no_dispatch()`, because I don't want the subclass to intercept my set_() call

(2) If I do that, the default behavior will be to dispatch to the cpu.cuda kernel for `set_()`. That's not actually what I want, since those two kernels don't support dynamic shapes (but the meta kernel does). That's why I swizzle the meta dispatch key here.

bdhirsh replied:
darn, good catch

bdhirsh replied:
Actually @albanD - I think Elias updated the `Meta` backend bit to be the highest priority backend bit ([code](https://github.com/pytorch/pytorch/blob/main/c10/core/DispatchKey.h#L49)).

This does mean though that backends that aren't part of `BackendComponent` might end up being higher priority:
- MkldnnCPU
- SparseCsrCPU/CUDA
- FPGA
- ORT
- Metal
- Vulkan


I wrote this function to do the meta swizzling the same way that it's done in fake tensor though, so I imagine that FakeTensor is equally broken with those backends. cc @eellison, what do you think?

For now, I can at least add a note here and reference it in fake tensor, so if/when we actually try to fix these backends later we'll know which code sites to audit.

bdhirsh replied:
To be fair though, most of these backends feel like "special" backends that aren't going to work with `torch.compile` for a whole host of other reasons anyway:
- metal/vulkan meant for mobile
- ORT having its own custom backend anyway
- SparseCSR can/should eventually be re-written as a functioanlity key

eellison replied:
That is correct, the other ones will not work currently. Why are the backends not part of `BackendComponent` ? If we added them to the local exclude set, would it work then ? 

I think i remember there being something weird about exclude/include, so maybe that's not viable..

bdhirsh replied:
>  Why are the backends not part of BackendComponent 

We have a pretty limited number of backend component slots (the "long term" solution is for more backends to use the out-of-tree PrivateUse1 registration. So we never ended up adding these extra backends here

> If we added them to the local exclude set, would it work then ? I think i remember there being something weird about exclude/include, so maybe that's not viable..

Actually, I think this should work. Although I'll leave it as not part of this PR

albanD replied:
Ho Elias is ahead of us haha!
Ignoring these non-generic backends sounds fine. cc @cpuhrsch FYI on the sparse one as it is the only one that may be used with compile.

bdhirsh committed:
Update on "add return_and_correct_aliasing() util for wrapper subclasses"

This PR adds a `return_and_correct_aliasing()` utility, that wrapper subclasses can use to get correct aliasing. I updated `TwoTensor` to use it, and added some testing that the aliasing of my `TwoTensor` subclass now matches the aliasing behavior of normal tensors.

Right now my test just uses a few hand-picked opinfos (that have varying aliasing behavior). I thought all op infos might be overkill (does that take a while to run?), but I'm happy to add them all if people prefer.

One more general question about this PR: eventually, proper aliasing will be a **requirement** in order for AOTAutograd to handle aliasing/mutations on subclasses properly during compilation. How can we make sure that wrapper subclasses use this API? A few options (from talking to Richard):

(1) Yolo require subclasses to use the API and hope users do as well (what this PR does)

(2) Yolo require subclasses to use the API, but add a kwarg to `_make_wrapper_subclass`, e.g. `manual_aliasing=True`, that torch.compile checks for before allowing the subclass to be used in compilation

(3) Automatically run this API in our python fallback, for **every** tensor subclass that currently implements `__tensor_flatten__` (aka only the "traceable" subclasses)

(4) Automatically run this API in our python fallback, for **every** tensor subclass. This would be a bit higher blast radius, since it would change the existing aliasing behavior of wrapper subclasses. Maybe.. this is the right thing to do though?

Either way, my tentative plan is to do (1) to unblock, and revisit this later once we want to come up with public docs + a more general "tensor subclass in PT2 requirements" plan




[ghstack-poisoned]

ezyang commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+            # For lists of outputs, need to alias every individual tensor to the input
+            alias_storage(args[0], o)
+
+def return_and_correct_aliasing(func, args, kwargs, out):
```
A top level comment explaining what this does and how you should comment would be greatly helpful!

bdhirsh replied:
will do!

ezyang commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+                alias_storage(args[arg_idx], outs[return_idx])
+
+    # Sigh... the torchscript parser has a bug where alias annotations for Tensor[](a) don't show up properly
+    # See https://github.com/pytorch/pytorch/issues/106173
```
IIRC, when @larryliu0820 ran into these problems, we opted to use the torchgen schema parser instead, re-parsing the schema from Python, which seemed to work a lot better.

bdhirsh replied:
Hmm, I could try that out (although tiny risk that using torchgen schema parsing on every op call through functionalization is a liiitle slow, maybe I should just do it for ops returning TensorList?)

ezyang commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+        mutated_args = [x for i, x in enumerate(args) if get_write_alias(func._schema.arguments[i]) is not None]
+        # Assumption: we have a very small number of inplace_view ops that follow a strict schema:
+        # there is only a single argument that gets its metadata mutated.
+        assert len(mutated_args) == 1
```
Is it actually a problem if you have more than one mutated arg? I don't see what would fail even if there were multiple

bdhirsh replied:
Yeah it shouldn't be a problem - so far I wrote this function in a more limited way that doesn't handle arbitrary schemas, and instead only handles the subset of schemas that I know exist in aten (for example, inplace_view ops always have a single mutated argument).

Happy to improve it as part of this PR, I just figured I wouldn't bother complicating it until someone complains.

albanD commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+            # For lists of outputs, need to alias every individual tensor to the input
+            alias_storage(args[0], o)
+
+def return_and_correct_aliasing(func, args, kwargs, out):
```
nit: not sure why "return" is in the name here?

bdhirsh replied:
to make it clear that "user subclasses should return the result of this function as the output of their torch_dispatch". Open to bikeshedding if you have any suggestions.

ezyang commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+        with torch.utils._mode_utils.no_dispatch():
+            func(*args, **kwargs)
+
+    # Next: we need to make sure to return inputs directly, if the output is a mutable alias (e.g. add_()).
```
Question: Wouldn't it be reasonable to expect the subclass to do this correctly? Unlike aliasing storages, it's pretty simple to make sure you return the correct input argument directly

bdhirsh replied:
So my thought here was that:

- if a subclass "works" with eager, we want to do all we can to mitigate silent correctness issues when they try to use it with torch.compile
- If we depend on the user to write their subclass to handle input mutation returns properly and they get it wrong, we're out of luck (they'll just be silently wrong in certain situations with torch.compile, where they happened to be ok in eager)

I agree that it's not **unreasonable** to ask the user to handle this, but offloading responsibility onto us seemed less risky. wdyt?

albanD replied:
I'm not sure why we're considering eager vs compile?
This is all about eager here.
The question, is are we ok with users writing inplace op to not return the input as-is. I agree with Ed that this is a pretty simple thing to do AND if they didn't do it, they most likely are not doing the inplace change properly.
So unless we have good examples showing that it is inconvenient for the subclass to return the input as-is, we shouldn't do it here.

ezyang commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
         transformed_tensors_dict[attr] = callback(attr, getattr(t, attr))
     return type(t).__tensor_unflatten__(transformed_tensors_dict, ctx)
+
+def _correct_storage_aliasing(func, args, outs):
```
High level description of what this does would be really helpful

ezyang commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+            # TODO: make this a context manager
+            torch._C._set_meta_in_tls_dispatch_include(True)
+            try:
+                torch.ops.aten.set_.source_Storage_storage_offset(ret, arg.untyped_storage(), 0, ret.shape)
```
Do you really want the offset to always be zero?

bdhirsh replied:
sorry thanks.... `ret.storage_offset()`

albanD commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+    # In the multi-return case, all aten ops return a tuple / list, so cast accordingly.
+    outs_to_return = type(out)([
+        args[get_arg_idx_from_alias(get_write_alias(func._schema.returns[i]))]
+        if get_write_alias(r) is not None else o
```
You know this is always not None here.

bdhirsh replied:
Hmm what about:
```
func: cummax.out(Tensor self, int dim, *, Tensor(a!) values, Tensor(b!) indices) -> (Tensor(a!) values, Tensor(b!) indices)
```

(multi-return, but both outputs are actually just the inputs).

Although I agree that as written, this logically is (unnecessarily) trying to account for an op that might have multiple returns, where some alias and some don't (which I don't think exists in aten today).

albanD replied:
> which I don't think exists in aten today

And you just checked on the line above that this is not the case here!

albanD commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+    if not all(get_write_alias(r) is not None for r in func._schema.returns):
+        raise RuntimeError("Unsupported schema: " + str(func._schema))
+
+    if len(func._schema.returns) == 1:
```
Why the rest below is needed? Why not assert that the user returns the input as-is properly? Having them create a new output just to drop it sounds weird?

bdhirsh replied:
hmm, I tried to answer this here: https://github.com/pytorch/pytorch/pull/107915/files#r1305804329, let me know what you think.

albanD commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+            # For lists of outputs, need to alias every individual tensor to the input
+            alias_storage(args[0], o)
+
+def return_and_correct_aliasing(func, args, kwargs, out):
```
This sounds ok. I wasn't expecting to see the inplace return in here: dropping the user outputs sounds a bit bad for me. Maybe make this an option? (can you share why you added that, I'm sure it's useful in your case above)
Also this function needs doc explaining what it does!

bdhirsh replied:
Definitely needs a doc, will add.

I answered the question on inplace returns in a comment above - lmk if you don't agree with it!
https://github.com/pytorch/pytorch/pull/107915#discussion_r1305804329

ezyang commented:
One thing that I'm conceptually having trouble with, is whether or not it is OK to do this if the inner tensors do not actually alias.

For example, suppose I have a wrapper subclass which has an inner tensor, and implements view() by cloning the inner tensor. return_and_correct_aliasing would happily common up the wrapper storage... even when there isn't actually any aliasing! (And vice versa.)

I suppose this is analogous to the debug autograd asserts problem, where autograd expects a lot of invariants to hold and it is easy for subclasses to break them. This PR makes it easier for the wrapper to report "accurate" aliasing info, but it does nothing to help implementers of subclasses actually implement correct semantics. So I am not really in favor of turning it on for everyone, you are talking the talk without walking the walk.

bdhirsh commented on test/test_python_dispatch.py:
```diff
@@ -2107,5 +2111,62 @@ def test_lstsq(self):
+    # so I picked a sampling of ops with representative schemas.
+    @ops([op for op in op_db if op.name in [
+        'mul',  # out-of-place
+        'cat',  # out-of-place (TensorList input)
```
fair point (i think they're in a separate db, I'll look)

bdhirsh commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+    flat_outs, _ = torch.utils._pytree.tree_flatten(outs)
+    for x in flat_outs:
+        if isinstance(x, torch.Tensor):
+            assert hasattr(x, '__torch_dispatch__')
```
I'm going to change this assert, to something like:

"When you call an aliasing op, the input and output must have the same subclass".

I don't think that this assert is.... strictly necessary, but it felt right (is there ever a case for views to change subclass-ness)? What do you think?

albanD replied:
As long as we're using as_strided to replay views, it is the right assert indeed.
If we were to use view_fn, we could relax that.
But I guess not that many people care, so this is ok for now.

bdhirsh commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+        if isinstance(x, torch.Tensor):
+            assert hasattr(x, '__torch_dispatch__')
+
+    def alias_storage(arg, ret):
```
yep that's right, will update

bdhirsh commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+        # there is only a single argument that gets its metadata mutated.
+        assert len(mutated_args) == 1
+        with torch.utils._mode_utils.no_dispatch():
+            func(*args, **kwargs)
```
you're right - will fix, thanks

bdhirsh commented on torch/utils/_python_dispatch.py:
```diff
@@ -164,3 +164,130 @@ def transform_subclass(t, callback):
+        with torch.utils._mode_utils.no_dispatch():
+            func(*args, **kwargs)
+
+    # Next: we need to make sure to return inputs directly, if the output is a mutable alias (e.g. add_()).
```
Yep, just via the inplace tag (which to be fair - functionalization already relies on this tag today to generate proper kernels for inplace views)

eellison reviewed (commented):
Is the intention to override incorrect meta functions? I dont know how we would properly model storage_offset() with this api.

bdhirsh committed:
Update on "add return_and_correct_aliasing() util for wrapper subclasses"

This PR adds a `return_and_correct_aliasing()` utility, that wrapper subclasses can use to get correct aliasing. I updated `TwoTensor` to use it, and added some testing that the aliasing of my `TwoTensor` subclass now matches the aliasing behavior of normal tensors.

Right now my test just uses a few hand-picked opinfos (that have varying aliasing behavior). I thought all op infos might be overkill (does that take a while to run?), but I'm happy to add them all if people prefer.

One more general question about this PR: eventually, proper aliasing will be a **requirement** in order for AOTAutograd to handle aliasing/mutations on subclasses properly during compilation. How can we make sure that wrapper subclasses use this API? A few options (from talking to Richard):

(1) Yolo require subclasses to use the API and hope users do as well (what this PR does)

(2) Yolo require subclasses to use the API, but add a kwarg to `_make_wrapper_subclass`, e.g. `manual_aliasing=True`, that torch.compile checks for before allowing the subclass to be used in compilation

(3) Automatically run this API in our python fallback, for **every** tensor subclass that currently implements `__tensor_flatten__` (aka only the "traceable" subclasses)

(4) Automatically run this API in our python fallback, for **every** tensor subclass. This would be a bit higher blast radius, since it would change the existing aliasing behavior of wrapper subclasses. Maybe.. this is the right thing to do though?

Either way, my tentative plan is to do (1) to unblock, and revisit this later once we want to come up with public docs + a more general "tensor subclass in PT2 requirements" plan




[ghstack-poisoned]

albanD commented:
> One thing that I'm conceptually having trouble with, is whether or not it is OK to do this if the inner tensors do not actually alias.

My feeling here is yes. As long as the Tensor behaves as-if it was aliased (any inplace op will reflect on the other Tensor).
For example, if you were to write a functionalization subclass, you can make view tracking as attributes on the subclass and make sure inplace are properly reflected without ever having the inner Tensor alias.

bdhirsh commented:
> Is the intention to override incorrect meta functions? I dont know how we would properly model storage_offset() with this api.

@eellison nope - the use case here is for something like:
```
@torch.compile
def f(x):
    x_view = x.view(-1)
    x_view.mul_(2)

x = MySubclass(....)
f(x)
```

In order for AOTAutograd to detect that there was an input mutation, the `x.view(-1)` needs to return a fresh (subclass) tensor that advertises as aliasing the storage of the input. The idea behind this PR is that users can easily get this wrong, so the PR provides a utility for subclass authors to use to make sure that they have correct aliasing.

bdhirsh committed:
Update on "add return_and_correct_aliasing() util for wrapper subclasses"

This PR adds a `return_and_correct_aliasing()` utility, that wrapper subclasses can use to get correct aliasing. I updated `TwoTensor` to use it, and added some testing that the aliasing of my `TwoTensor` subclass now matches the aliasing behavior of normal tensors.

Right now my test just uses a few hand-picked opinfos (that have varying aliasing behavior). I thought all op infos might be overkill (does that take a while to run?), but I'm happy to add them all if people prefer.

One more general question about this PR: eventually, proper aliasing will be a **requirement** in order for AOTAutograd to handle aliasing/mutations on subclasses properly during compilation. How can we make sure that wrapper subclasses use this API? A few options (from talking to Richard):

(1) Yolo require subclasses to use the API and hope users do as well (what this PR does)

(2) Yolo require subclasses to use the API, but add a kwarg to `_make_wrapper_subclass`, e.g. `manual_aliasing=True`, that torch.compile checks for before allowing the subclass to be used in compilation

(3) Automatically run this API in our python fallback, for **every** tensor subclass that currently implements `__tensor_flatten__` (aka only the "traceable" subclasses)

(4) Automatically run this API in our python fallback, for **every** tensor subclass. This would be a bit higher blast radius, since it would change the existing aliasing behavior of wrapper subclasses. Maybe.. this is the right thing to do though?

Either way, my tentative plan is to do (1) to unblock, and revisit this later once we want to come up with public docs + a more general "tensor subclass in PT2 requirements" plan




[ghstack-poisoned]

bdhirsh committed:
Update on "add return_and_correct_aliasing() util for wrapper subclasses"

This PR adds a `return_and_correct_aliasing()` utility, that wrapper subclasses can use to get correct aliasing. I updated `TwoTensor` to use it, and added some testing that the aliasing of my `TwoTensor` subclass now matches the aliasing behavior of normal tensors.

Right now my test just uses a few hand-picked opinfos (that have varying aliasing behavior). I thought all op infos might be overkill (does that take a while to run?), but I'm happy to add them all if people prefer.

One more general question about this PR: eventually, proper aliasing will be a **requirement** in order for AOTAutograd to handle aliasing/mutations on subclasses properly during compilation. How can we make sure that wrapper subclasses use this API? A few options (from talking to Richard):

(1) Yolo require subclasses to use the API and hope users do as well (what this PR does)

(2) Yolo require subclasses to use the API, but add a kwarg to `_make_wrapper_subclass`, e.g. `manual_aliasing=True`, that torch.compile checks for before allowing the subclass to be used in compilation

(3) Automatically run this API in our python fallback, for **every** tensor subclass that currently implements `__tensor_flatten__` (aka only the "traceable" subclasses)

(4) Automatically run this API in our python fallback, for **every** tensor subclass. This would be a bit higher blast radius, since it would change the existing aliasing behavior of wrapper subclasses. Maybe.. this is the right thing to do though?

Either way, my tentative plan is to do (1) to unblock, and revisit this later once we want to come up with public docs + a more general "tensor subclass in PT2 requirements" plan




[ghstack-poisoned]
